### ISSUE TITLE
Remove Attribute's IncludeInResult from Result and PolicyIssuer

### ISSUE THREAD

**Issue by cdanger** [oasis-tcs/xacml-spec#18]
The attribute `IncludeInResult` is always required on the `Attribute` element, even when returned in a `Result` or used in `PolicyIssuer` although it does not make sense there.
Is there a way to remove it in these cases? (But keep it in the Request.)
What are the options?

**Comment by steven-legg** [comment]
> The attribute `IncludeInResult` is always required on the `Attribute` element, even when returned in a `Result` where it does not make sense. Is there a way to remove it from the Result's Attributes? (But keep it in the Request.) What are the options?

Yes, `IncludeInResult` in a result is daft. It doesn't make sense in `PolicyIssuer` either. The JSON profile made it default false, which means it can be left out in a result. We could do the same in XSD.

Another option is to define a base `AttributeType` without `IncludeInResult` and define a `RequestAttributeType` with `IncludeInResult` derived from `AttributeType`. I'd still make `IncludeInResult` default false since I much less frequently need to set it to true. We would need to split `Attributes` as well (renamed to `Category` in the JSON profile).

**Comment by cdanger** [comment]
I prefer the second option, but in that case, we'd need a new element like `RequestAttribute` (of RequestAttributeType) as well, don't we?

So as I understand it, the changes to the XML schema if we go down that road:
- New AttributeType definition (without IncludeInResult):
```xml
<xs:element name="Attribute" type="xacml:AttributeType"/>
<xs:complexType name="AttributeType">
   <xs:sequence>
          <xs:element ref="xacml:AttributeValue" maxOccurs="unbounded"/>
   </xs:sequence>
   <xs:attribute name="AttributeId" type="xs:anyURI" use="required"/>
   <xs:attribute name="Issuer" type="xs:string" use="optional"/>
   <!-- IncludeInResult REMOVED -->
   <!--<xs:attribute name="IncludeInResult" type="xs:boolean" use="required"/>-->
</xs:complexType>
```

- New element/type `RequestAttribute` / `RequestAttributeType` (derived from the new AttributeType above, with optional IncludeInResult):
```xml
<!-- Other name suggestions besides RequestAttribute? -->
<xs:element name="RequestAttribute" type="xacml:AttributeType"/>
<xs:complexType name="RequestAttributeType">
   <xs:complexContent>
          <xs:extension base="xacml:AttributeType">
                <!-- IncludeInResult use changed from required to optional with default false -->
                <xs:attribute name="IncludeInResult" type="xs:boolean" use="optional" default="false"/>
          </xs:extension>
   </xs:complexContent>
</xs:complexType>
```

- New element/type `AttributeCategory`/`AttributeCategoryType` (or simply `Category`?) similar to `Attributes`/`AttributesType` but using the new `RequestAttribute` element above:
```xml
<xs:element name="AttributeCategory" type="xacml:AttributeCategoryType"/>
<xs:complexType name="AttributeCategoryType">
   <xs:sequence>
          <xs:element ref="xacml:Content" minOccurs="0"/>
          <xs:element ref="xacml:RequestAttribute" minOccurs="0"
              maxOccurs="unbounded"/>
   </xs:sequence>
   <!-- `Category` attribute may be redundant with the name AttributeCategory. So could we use 'name'... or ID? -->
   <xs:attribute name="Category" type="xs:anyURI" use="required"/>
   <xs:attribute ref="xml:id" use="optional"/>
</xs:complexType>
```

- `RequestType` changed to use the new AttributeCategory element above:
```xml
<xs:complexType name="RequestType">
   <xs:sequence>
          <xs:element ref="xacml:RequestDefaults" minOccurs="0"/>
          <!-- Changed Attributes to AttributeCategory -->
          <xs:element ref="xacml:AttributeCategory" maxOccurs="unbounded"/>
          <xs:element ref="xacml:MultiRequests" minOccurs="0"/>
   </xs:sequence>
   <xs:attribute name="ReturnPolicyIdList" type="xs:boolean" use="required"/>
   <xs:attribute name="CombinedDecision" type="xs:boolean" use="required" />
</xs:complexType>
```

**Comment by steven-legg** [comment]
> `Category` attribute may be redundant with the name AttributeCategory.

Yeah, it's a bit odd. How about `<Category>` for the element and `CategoryId` for the attribute? It aligns with `<Attribute>` and `AttributeId`. In that case we should also change the `Category` attribute to `CategoryId`in `<AttributeDesignator>` and `<AttributeSelector>` to be consistent.

**Comment by steven-legg** [comment]
`AttributeCategoryType` could be shortened to `CategoryType`.

**Comment by steven-legg** [comment]
These changes look reasonable.

A couple of things jump out at me. The Entity data type from the Entities Profile has an indicative XML Schema definition, that is effectively a base type for CategoryInRequestType:
```
<xs:complexType name="EntityType">
  <xs:sequence>
    <xs:element ref="xacml:Content" minOccurs="0"/>
    <xs:element ref="xacml:Attribute" minOccurs="0" maxOccurs="unbounded"/>
  </xs:sequence>
</xs:complexType>
```
The xml:id XML attribute should become an unqualified attribute for the sake of the JSON and YAML representations. It shouldn't be called "id" if we also have "Id" and shouldn't have "xml" in its name.

**Comment by cdanger** [comment]
> These changes look reasonable.
> 
> A couple of things jump out at me. The Entity data type from the Entities Profile has an indicative XML Schema definition, that is effectively a base type for CategoryInRequestType:
> 
> ```
> <xs:complexType name="EntityType">
>   <xs:sequence>
>     <xs:element ref="xacml:Content" minOccurs="0"/>
>     <xs:element ref="xacml:Attribute" minOccurs="0" maxOccurs="unbounded"/>
>   </xs:sequence>
> </xs:complexType>
> ```

Interesting point, I forgot about that one, especially as [we agreed to add it to the core schema](#21). 
**Do you need the IncludeInResult actually in a EntityType's Attribute?**  In which case it's a bit different and I should take it into account in my schema change proposal (Attribute element).

> 
> The xml:id XML attribute should become an unqualified attribute for the sake of the JSON and YAML representations. It shouldn't be called "id" if we also have "Id" and shouldn't have "xml" in its name.

In that case, I prefer to rename "Id" back to "CategoryId", and use "Id" as replacement for "xml:id".

For the record and also to emphasize the implications of not using "xml:id" anymore: "xml:id" is not just any ID attribute but a standard XML ID attribute defined in the [[XmlId] standard from W3C](https://www.w3.org/TR/xml-id/#intro) as mentioned in XACML 3.0 core spec, with special semantics and processing rules. Section [D.2 (With XML schema validation)](https://www.w3.org/TR/xml-id/#with-schema-validation) says in particular: **_XML Schema authors are encouraged to use xml:id attributes when providing identifiers for elements declared in their schemas._** (Identifier here means a it is unique across the XML document.) Therefore, if we remove "xml:id" from the XML, we choose to ignore this recommendation in fact. Just to be aware.

That being said, I don't mind the change, as "xml:id" is used only by section 3.4 of the Multiple Decision profile (By reference to <Attributes> elements) and my implementation does not support this (and no plan to do so) at the moment. Also I looked at another well-known XML standard from W3C - XMLDSig - where they use a bunch of ID and ID reference, and they don't use "xml:id", just "Id" (with type xs:ID, and the references must have type xs:IDREF), although XMLDSig 1.1 for instance is much more recent than the xml:id standard. So there is at least one well-known example that does not follow this, even in W3C.

**Comment by cdanger** [comment]
See commit  f82e720 for final changes
